File: 20250509/도넛과_막대_그래프.java
import java.util.*;

class Solution {
    public int[] solution(int[][] edges) {
        int[] answer = new int[4];

        Map<Integer, List<Integer>> graph = new HashMap<>();
        Map<Integer, Integer> startCounts = new HashMap<>();
        Map<Integer, Integer> endCounts = new HashMap<>();

        // - 그래프를 만들면서 각 정점에 연결된 간선 수를 파악
        for (int[] edge : edges) {
            int start = edge[0];
            int end = edge[1];

            // 그래프 세팅
            graph.putIfAbsent(start, new ArrayList<>());
            graph.get(start).add(end);

            startCounts.put(start, startCounts.getOrDefault(start, 0) + 1);
            endCounts.put(end, endCounts.getOrDefault(end, 0) + 1);
        }

        // 한 번도 end로 값이 안들어온 정점이 새로 만든 정점 k
        for (Integer key : startCounts.keySet()) {
            if (startCounts.get(key) > 1 && !endCounts.containsKey(key) ) {
                answer[0] = key;
                break;
            }
        }

        // - k 노드에서 각 노드 탐색 시작
        for (Integer i : graph.get(answer[0])) {
            // 정점에 연결된 간선이 1개 초과이면? 8자
            // start 정점으로 돌아왔으면? 도넛
            // 더이상 갈 정점이 없으면? 막대
            GraphType graphType = getGraphType(graph, i);
            if (graphType == GraphType.DONUT) answer[1]++;
            if (graphType == GraphType.STRAIGHT) answer[2]++;
            if (graphType == GraphType.Eights) answer[3]++;
        }

        return answer;
    }

    GraphType getGraphType(Map<Integer, List<Integer>> graph, int node) {
        int i = node;
        while (true) {
            List<Integer> nodes = graph.get(i);
            if (nodes == null) return GraphType.STRAIGHT;

            if (nodes.size() > 1) return GraphType.Eights;

            i = nodes.get(0);
            if (i == node) return GraphType.DONUT;
        }
    }

    enum GraphType {
        DONUT,
        STRAIGHT,
        Eights,
    }
}

File: 20250509/도넛과_막대_그래프.js
function solution(edges) {
    const inOutMap = new Map();

    // 1. 각 노드의 out-degree, in-degree 계산
    for (const [from, to] of edges) {
        // 각 노드에 대해 [out-degree, in-degree]를 저장
        if (!inOutMap.has(from)) inOutMap.set(from, [0, 0]);
        if (!inOutMap.has(to)) inOutMap.set(to, [0, 0]);

        // from은 나가는 간선이므로 out-degree 증가
        inOutMap.get(from)[0] += 1;
        // to는 들어오는 간선이므로 in-degree 증가
        inOutMap.get(to)[1] += 1;
    }

    let createdNode = 0; // 생성된 중앙 노드
    let donut = 0;       // 도넛 그래프 수
    let stick = 0;       // 막대 그래프 수
    let eight = 0;       // 8자 그래프 수

    // 2. 각 노드를 순회하면서 생성 정점과 그래프 유형을 분류
    for (const [node, [out, inn]] of inOutMap.entries()) {
        if (out >= 2 && inn === 0) {
            // 생성 정점은: 나가는 간선이 2개 이상, 들어오는 간선이 없음
            createdNode = node;
        } else if (out === 0) {
            // 막대 그래프의 끝 노드는 나가는 간선이 없음
            stick += 1;
        } else if (out >= 2 && inn >= 2) {
            // 8자 그래프의 중심 노드는 in/out이 모두 2 이상
            eight += 1;
        }
    }

    // 3. 도넛 그래프 수 계산
    // 생성 정점에서 나간 총 간선 수 - 막대 수 - 8자 수
    donut = (inOutMap.get(createdNode)?.[0] || 0) - stick - eight;

    return [createdNode, donut, stick, eight];
}
